README.md
template_block
=====

The 'template block' is a template engine for the code generators with a very simple syntax.

Version: 0.1.6

```dart
import "package:template_block/template_block.dart";

void main() {
  var statement = new TemplateBlock(_templateStatement);
  statement.assign("MIN", 0);
  statement.assign("MAX", 128);
  var methodFoo = new TemplateBlock(_templateMethodFoo);
  methodFoo.assign("#STATEMENTS", statement.process());
  statement = statement.clone();
  statement.assign("MIN", 512);
  statement.assign("MAX", 1024);
  methodFoo.assign("#STATEMENTS", statement.process());
  var methodBaz = new TemplateBlock(_templateMethodBaz);
  var clazz = new TemplateBlock(_templateClass);
  clazz.assign("NAME", "MyClass");
  clazz.assign("#METHODS", methodBaz.process());
  clazz.assign("#METHODS", methodFoo.process());
  var library = new TemplateBlock(_templateLibraryPart);
  library.assign("NAME", "my_lib");
  library.assign("#CLASSES", clazz.process());
  print(library.process().join("\n"));
}

final String _templateLibraryPart =
    '''
part of {{NAME}};

{{#CONSTANTS}}
{{#VARIABLES}}
{{#CLASSES}}
''';

final String _templateClass =
    '''
class {{NAME}} {
  {{#CONSTANTS}}
  {{#VARIABLES}}
  {{#METHODS}}
}
''';

final String _templateMethodFoo =
    '''
int foo(int start, int end) {
  $BAZ_NAME();
  {{#STATEMENTS}}
}
''';

const String BAZ_NAME = "baz";

final String _templateMethodBaz = '''
void $BAZ_NAME() {
}
''';

final String _templateStatement =
    '''
for (var i = {{MIN}}; i < {{MAX}}; i++) {
}''';

```
**Output:**

```dart
part of my_lib;

class MyClass {
  void baz() {
  }
  
  int foo(int start, int end) {
    baz();
    for (var i = 0; i < 128; i++) {
    }
    for (var i = 512; i < 1024; i++) {
    }
  }
  
}

```

**Grammar**

```
template <- ((lastLine / multiLine / singleLine / emptyLine) (NEW_LINE !EOF)?)* EOF

lastLine <- NEW_LINE &EOF

multiLine <- multiLinePrefix* multiLineKey multiLineSuffix*

multiLinePrefix <- !multiLineKey !NEW_LINE .

multiLineKey <- "{{#" IDENT "}}"

multiLineSuffix <- !NEW_LINE .

singleLine <- singleLineParts

singleLineParts <- (singleLineKey / singleLinePart)+

singleLineKey <- "{{" IDENT "}}"

singleLinePart <- !singleLineKey !NEW_LINE .

emptyLine <- &NEW_LINE

EOF <- !.

IDENT <- IDENT_START IDENT_CONT*

IDENT_START <- [A-Za-z] / "_"

IDENT_CONT <- [0-9A-Za-z] / "_"

NEW_LINE <- "\n" / "\r\n" / "\r"
```

CHANGELOG.md
## 0.2.0

- Adaptation to Dart 2.2

## 0.1.6

- Fixed bug template parser

## 0.1.5

- Generated the new parser with using the new version of `peg` parser generator.

## 0.1.3

- Generated the new parser according to changes in `strings` packages

## 0.1.2

- Fixed bug in grammar (character `_` added to the allowed part of `IDENT`)

## 0.1.1

- Template parser has been rewritten using PEG grammar



tool/project.dart
import "dart:io";
import "package:build_tools/build_shell.dart";
import "package:build_tools/build_tools.dart";
import "package:build_tools/build_utils.dart";
import "package:file_utils/file_utils.dart";

const String CHANGE_LOG = "tool/change.log";
const String CHANGELOG_MD = "CHANGELOG.md";
const String EXAMPLE_DART = "example/example.dart";
const String PUBSPEC_YAML = "pubspec.yaml";
const String README_MD = "README.md";
const String README_MD_IN = "tool/README.md.in";

void main(List<String> args) {
  // http://dartbug.com/20119 (before change directory)
  var script = Platform.script;

  // Change directory to root
  FileUtils.chdir("..");

  file(CHANGELOG_MD, [CHANGE_LOG], (Target t, Map args) {
    writeChangelogMd();
  });

  file(README_MD_IN, [], (Target t, Map args) {
    FileUtils.touch([t.name], create: true);
  });

  file(README_MD, [README_MD_IN, PUBSPEC_YAML, EXAMPLE_DART], (Target t, Map
      args) {
    var sources = t.sources.toList();
    var template = new File(sources.removeAt(0)).readAsStringSync();
    // Remove "pubspec.yaml"
    sources.removeAt(0);
    for (var filename in sources) {
      var text = new File(filename).readAsStringSync();
      template = template.replaceFirst("{{$filename}}", text);
    }

    template = template.replaceFirst("{{DESCRIPTION}}", getDescription());
    template = template.replaceFirst("{{VERSION}}", getVersion());
    new File(t.name).writeAsStringSync(template);
  });

  target("default", ["git:status"], null, description: "git status");

  target("git:status", [], (Target t, Map args) {
    return exec("git", ["status", "--short"]);
  }, description: "git status --short");

  target("git:add", [], (Target t, Map args) {
    return exec("git", ["add", "--all"]);
  }, description: "git add --all");

  target("git:commit", [CHANGELOG_MD, README_MD, "git:add"], (Target t, Map
      args) {
    var message = args["m"];
    if (message == null || message.isEmpty) {
      print("Please, specify the `commit` message with --m option");
      return -1;
    }

    return exec("git", ["commit", "-m", message]);
  }, description: "git commit, --m \"message\"");

  after(["git:commit"], (Target t, Map args) {
    var version = incrementVersion(getVersion());
    print("Change the project version to '$version' (Y/N)?");
    if (stdin.readLineSync().toLowerCase().startsWith("y")) {
      updateVersion(version);
      print("Version switched to $version");
    }
  });

  target("git:push", [], (Target t, Map args) {
    // TODO: The `exec git push` does not show the prompt on Windows.
    // But on Posix everything works as expected.
    // Problem in the "git" or in the Dart VM "stdxxx" streams.
    return exec("git", ["push", "origin", "master"]);
  }, description: "git push origin master");

  target("log:changes", [], (Target t, Map args) {
    var message = args["m"];
    if (message == null || message.isEmpty) {
      print("Please, specify the `message` with --m option");
      return -1;
    }

    logChanges(message);
  }, description: "log changes, --m message", reusable: true);

  target("prj:changelog", [CHANGELOG_MD], null, description:
      "generate '$CHANGELOG_MD'", reusable: true);

  target("prj:readme", [README_MD], null, description: "generate '$README_MD'",
      reusable: true);

  target("prj:version", [], (Target t, Map args) {
    print("Version: ${getVersion()}");
  }, description: "display version", reusable: true);

  new BuildShell().run(args).then((exitCode) => exit(exitCode));
}

String getDescription() {
  var file = new File(PUBSPEC_YAML);
  var lines = file.readAsLinesSync();
  var description = "";
  for (var line in lines) {
    if (line.startsWith("description")) {
      var index = line.indexOf(":");
      if (index != -1 && line.length > index + 1) {
        description = line.substring(index + 1).trim();
      }
    }
  }

  return description;
}

String getVersion() {
  var file = new File(PUBSPEC_YAML);
  var lines = file.readAsLinesSync();
  var version = "0.0.1";
  for (var line in lines) {
    if (line.startsWith("version")) {
      var index = line.indexOf(":");
      if (index != -1 && line.length > index + 1) {
        version = line.substring(index + 1).trim();
      }
    }
  }

  return version;
}

String incrementVersion(String version) {
  var parts = version.split(".");
  if (parts.length < 3) {
    return version;
  }

  var patch = int.parse(parts[2], onError: (x) => null);
  if (patch == null) {
    return version;
  }

  parts[2] = ++patch;
  parts.length = 3;
  return parts.join(".");
}

void logChanges(String message) {
  if (message == null || message.isEmpty) {
    return;
  }

  FileUtils.touch([CHANGE_LOG], create: true);
  var file = new File(CHANGE_LOG);
  var length = file.lengthSync();
  var fp = file.openSync(mode: FileMode.APPEND);
  var sb = new StringBuffer();
  if (length != 0) {
    sb.writeln();
  }

  sb.write("${getVersion()} $message");
  fp.writeStringSync(sb.toString());
  fp.closeSync();
}

void updateVersion(String version) {
  var file = new File(PUBSPEC_YAML);
  var found = false;
  var lines = file.readAsLinesSync();
  var sb = new StringBuffer();
  for (var line in lines) {
    if (line.startsWith("version")) {
      found = true;
      break;
    }
  }

  if (!found) {
    var pos = lines.length == 0 ? 0 : 1;
    lines.insert(pos, "version: $version");
  }

  for (var line in lines) {
    if (line.startsWith("version")) {
      sb.writeln("version: $version");
    } else {
      sb.writeln(line);
    }
  }

  var string = sb.toString();
  file.writeAsStringSync(string);
}

void writeChangelogMd() {
  FileUtils.touch([CHANGELOG_MD], create: true);
  var log = new File(CHANGE_LOG);
  if (!log.existsSync()) {
    return;
  }

  var lines = log.readAsLinesSync();
  lines = lines.reversed.toList();
  var versions = <String, List<String>> {};
  for (var line in lines) {
    var index = line.indexOf(" ");
    if (index != -1) {
      var version = line.substring(0, index);
      var message = line.substring(index + 1).trimLeft();
      var messages = versions[version];
      if (messages == null) {
        messages = <String>[];
        versions[version] = messages;
      }

      messages.add(message);
    }
  }

  var sb = new StringBuffer();
  for (var version in versions.keys) {
    sb.writeln("## ${version}");
    sb.writeln();
    var messages = versions[version];
    messages.sort((a, b) => a.compareTo(b));
    for (var message in messages) {
      sb.writeln("- $message");
    }

    sb.writeln();
  }

  var md = new File(CHANGELOG_MD);
  md.writeAsStringSync(sb.toString());
}

example/example.dart
import "package:template_block/template_block.dart";

void main() {
  var statement = TemplateBlock(_templateStatement);
  statement.assign("MIN", 0);
  statement.assign("MAX", 128);
  var methodFoo = TemplateBlock(_templateMethodFoo);
  methodFoo.assign("#STATEMENTS", statement.process());
  statement = statement.clone();
  statement.assign("MIN", 512);
  statement.assign("MAX", 1024);
  methodFoo.assign("#STATEMENTS", statement.process());
  var methodBaz = TemplateBlock(_templateMethodBaz);
  var clazz = TemplateBlock(_templateClass);
  clazz.assign("NAME", "MyClass");
  clazz.assign("#METHODS", methodBaz.process());
  clazz.assign("#METHODS", methodFoo.process());
  var library = TemplateBlock(_templateLibraryPart);
  library.assign("NAME", "my_lib");
  library.assign("#CLASSES", clazz.process());
  print(library.process().join("\n"));
}

final String _templateLibraryPart = '''
part of {{NAME}};

{{#CONSTANTS}}
{{#VARIABLES}}
{{#CLASSES}}
''';

final String _templateClass = '''
class {{NAME}} {
  {{#CONSTANTS}}
  {{#VARIABLES}}
  {{#METHODS}}
}
''';

final String _templateMethodFoo = '''
int foo(int start, int end) {
  $BAZ_NAME();
  {{#STATEMENTS}}
}
''';

const String BAZ_NAME = "baz";

final String _templateMethodBaz = '''
void $BAZ_NAME() {
}
''';

final String _templateStatement = '''
for (var i = {{MIN}}; i < {{MAX}}; i++) {
}''';
